src/Analysis/Ast/Impl/Analyzer/PythonAnalyzer.cs

```diff
@@ -14,145 +14,366 @@
+            public void TrySetException(Exception ex, int version) {
+                if (version >= Version) {
+                    _analysisTcs.TrySetException(ex);
+                }
+            }
 
-                foreach (var c in node.Children) {
-                    await AnalyzeChainAsync(c, cts.Token);
+            public void TryCancel(OperationCanceledException oce, int version) {
+                if (version >= Version) {
+                    _analysisTcs.TrySetCanceled(oce.CancellationToken);
                 }
             }
-        }
 
-        private static void NotifyAnalysisComplete(IDependencyChainNode node, IDocumentAnalysis analysis) {
-            if (!node.Analyzable.NotifyAnalysisComplete(analysis)) {
-                // If snapshot does not match, there is no reason to continue analysis along the chain
-                // since subsequent change that incremented the expected version will start
-                // another analysis run.
-                throw new OperationCanceledException();
+            public void Invalidate(int version) {
+                if (Version >= version) {
+                    return;
+                }
+                Version = version;
+
+                if (!_analysisTcs.TrySetCanceled() && _analysisTcs.Task.Status == TaskStatus.RanToCompletion) {
+                    PreviousAnalysis = _analysisTcs.Task.Result;
+                }
+                _analysisTcs = new TaskCompletionSource<IDocumentAnalysis>();
             }
         }
 
-        /// <summary>
-        /// Performs analysis of the document. Returns document global scope
-        /// with declared variables and inner scopes. Does not analyze chain
-        /// of dependencies, it is intended for the single file analysis.
-        /// </summary>
-        private async Task<IDocumentAnalysis> AnalyzeAsync(IDependencyChainNode node, CancellationToken cancellationToken) {
-            var startTime = DateTime.Now;
+        private sealed class ModuleDependencyFinder : IDependencyFinder<ModuleKey, AnalysisEntry> {
+            private readonly IFileSystem _fileSystem;
 
-            //_log?.Log(TraceEventType.Verbose, $"Analysis begins: {node.Document.Name}({node.Document.ModuleType})");
-            // Store current expected version so we can see if it still 
-            // the same at the time the analysis completes.
-            var analysisVersion = node.Analyzable.ExpectedAnalysisVersion;
+            public ModuleDependencyFinder(IFileSystem fileSystem) {
+                _fileSystem = fileSystem;
+            }
+
+            public Task<ImmutableArray<ModuleKey>> FindDependenciesAsync(AnalysisEntry value, CancellationToken cancellationToken) {
+                var dependencies = new HashSet<ModuleKey>();
+                var module = value.Module;
+                var isTypeshed = module.ModuleType == ModuleType.Stub; // TODO: This is not the correct way to determine Typeshed.
```

Comment:
I think `ModuleKey` has `IsTypeshed`. Or maybe `ModuleKey` could just hold the desired resolution service instance.

Reply:
Sure it can, but it doesn't solve the question how to determine which resolution service instance is the right one.

Reply:
I guess when module is first created as a stub we can attach the stub resolution service to it so when we need to find dependencies we'll just use the service attached to the module.

Reply:
Good idea, will do.